### PR TITLE
Linear KF without covariance transport

### PR DESCRIPTION
**Pull request by VanillaBrooks**

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+#[derive(Debug)]
+pub enum MatrixError {
+    NonInvertible,
+}
+
+// this function is only here to ensure that all `std::From` trait implementations 
+// are correctly expanded at compile time. It never needs to be called
+#[allow(dead_code)]
+fn init_from() {
```
Comment (paulgessinger) [inline src/error.rs:62]: So this makes the macro actually expand for each type? Interesting that this works inside a function...

File: src/filter/filter_gain.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
+use nalgebra as na;
+use super::super::config::*;
+
+// pub fn measurement_vector(
+//     jacobian: &Mat5,
+//     previous_state_vec: &Vec5) -> Vec2{
+//     return jacobian * previous_state_vec
+// }
+
+pub fn state_vector( // x
```
Comment (paulgessinger) [inline src/filter/filter_gain.rs]: This is the *filtering* step, and `extrap_state_vector` below is the *predicted* state. Maybe it makes sense to note that in a comment.

File: src/filter/filter_gain.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    measurement : &Vec2, 
+    sensor_mapping: &Mat2x5) -> Vec5 {
+
+    let parens = measurement - (sensor_mapping * extrap_state_vector);
+    let kalman_product = kalman_gain * parens;
+    return extrap_state_vector + kalman_product;
+}
+
+//TODO: remove `unwrap` on the inverse
+#[allow(dead_code)]
```
Comment (paulgessinger) [inline src/filter/filter_gain.rs]: Why is this needed here?

File: src/filter/filter_gain.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+pub fn kalman_gain ( // K
+    C : &Mat5,
+    H : &Mat2x5,
```
Comment (paulgessinger) [inline src/filter/filter_gain.rs]: This is the *measurement mapping function*. I think eventually it makes sense to call it the same consistently.

File: src/filter/filter_gain.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    V : &Mat2) -> Mat5x2 {
+
+    let parens = V + ( H * C * H.transpose() );
+    let kalman_gain = C * H.transpose() * parens.try_inverse().unwrap();
+    
+    kalman_gain
+}
+
+//TODO add lazy static for identity
```
Comment (paulgessinger) [inline src/filter/filter_gain.rs]: Would that be thread-safe? I don't know the details of `nalgebra`'s implementation, but I'd imagine `identity` is known at compile time and optimized.

File: src/filter/filter_gain.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+//TODO  add second method for this calculation
+//      since the second one looks to have less matmul
+
+//TODO add lazy static for identity
+pub fn residual_vec(  //r
+    H : &Mat2x5,
+    K : &Mat5x2,
+    residual_preiction : &Vec2) -> Vec2 {
```
Comment (paulgessinger) [inline src/filter/filter_gain.rs]:
```suggestion
    residual_prediction : &Vec2) -> Vec2 {
```

File: src/filter/filter_gain.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    let ident = Mat2::identity();
+    let parens = ident - (H * K);
+
+    return  parens * residual_preiction;
+}
+
+pub fn residual_mat( //R
```
Comment (paulgessinger) [inline src/filter/filter_gain.rs:57]: Not sure if you'd call this a *residual*.

File: src/filter/filter_gain.rs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+//     return jacobian * previous_state_vec
+// }
+
+pub fn state_vector( // x
+    extrap_state_vector: &Vec5, 
+    kalman_gain: &Mat5x2, 
+    measurement : &Vec2, 
+    sensor_mapping: &Mat2x5) -> Vec5 {
+
+    let parens = measurement - (sensor_mapping * extrap_state_vector);
```
Comment (paulgessinger) [inline src/filter/filter_gain.rs]: I guess you could use `residual_vec` from below.

File: src/filter/filter_means.rs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    
+    let product_one = C_prediction.try_inverse().unwrap() * extrap_state_vector;
+    let product_two = H.transpose() * G * m_k;
+
+    return C_k * (product_one + product_two)
+}
+
+//TODO : fix unwrap
+#[allow(dead_code)]
+fn covariance_mat ( // C_k
```
Comment (paulgessinger) [inline src/filter/filter_means.rs]: In `filter_gain.rs` you call this `covariance_matrix`.

File: src/filter/linear.rs [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+)  -> SmoothedData{
+    
+    if (V_vec.len() == H_vec.len()) && (H_vec.len() == m_k_vec.len()) {}
+    else {
+        panic!("vector lengths need to be the same length")
+    } 
+    let input_length = V_vec.len();
+
+    store_vec!{
+        input_length + 1; // length all vectors will be initialized to (+! for seeded)
```
Comment (paulgessinger) [inline src/filter/linear.rs]:
```suggestion
        input_length + 1; // length all vectors will be initialized to (+1 for seeded)
```

File: src/filter/utils.rs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+/// Placeholder function for some form of effective seeding of Mat5s
+pub fn seed_state_vec() -> Vec5 {
+    return Vec5::new_random_generic(na::U5, na::U1)
+}
+
+/// Creates a vector of `num` length with Mat5 components
+pub fn vec_of_mat(num: usize) -> Vec<Mat5> {
+    
+    let mut return_vec: Vec<Mat5> = Vec::with_capacity(num);
+    (0..num).into_iter()
```
Comment (paulgessinger) [inline src/filter/utils.rs:19]: Not sure this makes sense. You're initializing all the covariances to the seed value (and below the same for the state vectors). But what we "seed" here is only the initial direction, and an initial guess of the uncertainty.

File: src/filter/prediction.rs [hunk 12]
```diff
@@ -0,0 +1,29 @@
+use nalgebra as na;
+use super::super::config::*;
+
+// extrapolating state vector
+pub fn state_vector (jacobian: &Mat5, state_vector: &Vec5) -> Vec5 {
+    return jacobian * state_vector
```
Comment (paulgessinger) [inline src/filter/prediction.rs]: This works for the linear case (straight line), but will need to change later probably.

File: src/filter/prediction.rs [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+// extrapolation of the covariance matrix 
+pub fn covariance_matrix(jacobian: &Mat5, previous_covariance: &Mat5)-> Mat5{
+    return jacobian * previous_covariance * jacobian.transpose()
+}
+
+// just below eq. 7
+// covariance of predicted results
+pub fn residual_mat(V: &Mat2, H: &Mat2x5, C: &Mat5) -> Mat2 {
```
Comment (paulgessinger) [inline src/filter/prediction.rs]: This is duplicated above somewhere. Intentionally?

File: src/filter/prediction.rs [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    return V + (H*C * H.transpose())
+}
+
+pub fn residual_vec(
```
Comment (paulgessinger) [inline src/filter/prediction.rs:35]: Same here.

File: src/filter/macros.rs [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+#[macro_export]
+/// Get a reference to an item in a vector using .get_unchecked(_)
+macro_rules! get_unchecked {
+    ($index:expr; $($vector:ident => $destination:ident),+) => {
+        $(
+            let $destination = 
+                unsafe {
+                    $vector.get($index).expect("get_unchecked!{} fetched something out of bounds")
```
Comment (paulgessinger) [inline src/filter/macros.rs:51]: You're not actually using `get_unchecked` here. And does `get_unchecked` return an enum? I guess the point of *unchecked* is that it doesn't?

File: src/filter/macros.rs [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+             let coll: Vec<_> = $var.collect();
+             dbg!{coll.len()};
+         )+
+    };
+}
+
+
+#[macro_export]
+/// Used for getting the next value of the iterator. Additionally, this macro allows
+/// us to move the ownership of the data to "stagger" it. This does the following:
```
Comment (paulgessinger) [inline src/filter/macros.rs:71]: So by stagger you mean basically that you iterate pairwise, correct?

Comment (paulgessinger) [reply src/filter/macros.rs:71]: I'm not entirely sure how much you want this behaviour hidden behind macros, since the way the iteration works is pretty central to the entire algorithm. What do you think?

File: src/filter/linear.rs [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+use super::super::geometry::traits::{Plane, Transform};
+
+use super::utils::SmoothedData;
+
+#[macro_use]
+use super::macros;
+
+/// Monolithic function to handle linear KF calculations
+#[allow(dead_code)]
+pub fn run(
```
Comment (paulgessinger) [inline src/filter/linear.rs:20]: Can you add a comment what's supposed to be what here?

File: src/filter/linear.rs [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+    V_vec: &Vec<Mat2>, 
+    H_vec: &Vec<Mat2x5>,
```
Comment (paulgessinger) [inline src/filter/linear.rs]: If we're only ever going to use 2D plane measurements, then this will always be the exact same projection matrix
```
1 0 0 0 0 
0 1 0 0 0
```

One could potentially remove the input vector here, I guess.

File: src/filter/linear.rs [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+    // Store the seeded values in their respective iterators
+    push!(
+        previous_covariance => filter_cov_mat_iter, 
+        previous_state_vec => filter_state_vec_iter
+    );
+
+    let mut H_iter = H_vec.iter();
+    let mut V_iter = V_vec.iter();
+    let mut M_k_iter = m_k_vec.iter();
```
Comment (paulgessinger) [inline src/filter/linear.rs]: So the input measurements are all stored in a vector, and you iterate forward in lockstep. What happens in case the prediction passes through a sensor in the geometry, but it doesn't find a corresponding measurement? We call that a whole, and in the real world, it can happen. Any idea on how you would handle this?

File: src/geometry/rectangle.rs [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+    to_global: Aff3,    // L => G for point
+    to_local: Aff3,     // G => L for point
+    measurement_to_state_vector: Mat2x5
+}
+
+impl Rectangle {
+    /*
+    /// This is the constructor for the rectangular geometry. It expects a 4x4 `nalgebra::Matrix4<f32>` that is invertible 
+    /// and a 4 element array of `nalgebra::Point3<f32>`. If the matrix is not invertible it will return `Err(&str)`.
+    /// The provided matrix should be an affine transformation for converting from R2->R3
```
Comment (paulgessinger) [inline src/geometry/rectangle.rs]: Nitpick: the affine transform transforms from R3 -> R3, it's just that we then choose to ignore the z coordinate, thus effectively turning it into R2.

File: src/geometry/rectangle.rs [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+    ///let base_len = 3.0;
+    ///let height_len = 3.0;
+    ///let tfm_matrix= na::Matrix4::<f64>::identity(); //arbitrary transform matrix
+    ///let projection = na::Matrix5x2::<f64>::zeros();
+    ///Rectangle::new(base_len, height_len, tfm_matrix, projection);
+    /// 
+    /// 
+    /// ```
+    /// */
+    pub fn new(base: Real, height: Real, to_global_tfm_matrix: Mat4, projection_mat: Mat2x5) -> Result<Rectangle, MatrixError>{
```
Comment (paulgessinger) [inline src/geometry/rectangle.rs]: If we assume 2D sensors, you could go with the assumption that after the transformation is done, the first two coordinates are the local coordinates. That way, the projection matrix is always the same.

File: src/geometry/rectangle.rs [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+    /// use kalman_rs::geometry::rectangle::Rectangle;
+    /// 
+    /// let base = 3.0;
+    /// let height = 3.0;
+    /// let tfm_matrix : na::Matrix4<f64>= na::Matrix4::new(1.0,0.0,0.0,0.0,  0.0,1.0,0.0,0.0,  0.0,0.0,1.0,0.0, 0.0,0.0,0.0,1.0);
+    /// let mut rectangle_sensor = Rectangle::new(base, height, tfm_matrix).unwrap();
+    /// 
+    /// let is_point_on_sensor = rectangle_sensor.contains_from_local(&na::Point2::new(1.0, 6.0));
+    /// ```*/
+    fn contains_from_local(&self, input: &P2) -> bool {
```
Comment (paulgessinger) [inline src/geometry/rectangle.rs]: Maybe just call it `inside()`?

File: src/geometry/rectangle.rs [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+        
+        if (input.x.abs() < self.half_base.abs()) && (input.y.abs() < self.half_height.abs()) {
```
Comment (paulgessinger) [inline src/geometry/rectangle.rs:145]: I guess you can return the boolean directly :smile:

File: src/geometry/rectangle.rs [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+    /// let height = 3.0;
+    /// let tfm_matrix : na::Matrix4<f64>= na::Matrix4::new(1.0,0.0,0.0,0.0,  0.0,1.0,0.0,0.0,  0.0,0.0,1.0,0.0, 0.0,0.0,0.0,1.0);
+    /// let mut rectangle_sensor = Rectangle::new(base, height, tfm_matrix).unwrap();
+    /// 
+    /// let on_plane = rectangle_sensor.on_plane(&na::Point3::new(1.0, 3.0, 0.0)); //true
+    /// ```*/
+    fn on_plane(&self, input_point: &P3) -> bool {
+        let pv : Vec3= P3::new(0.0, 0.0, 0.0) - input_point;
+        //TODO : this function should probably not return result
+        if self.normal.dot(&pv) == 0.0 {
```
Comment (paulgessinger) [inline src/geometry/rectangle.rs]: Exactly zero is probably problematic, because it's a floating point operation. You should check if it is below a certain small epsilon > 0.

**Review by paulgessinger — COMMENTED** [review]
Looks good! I'll try to add some basic travis ci config to test if the code compiles and the tests pass (but on a separate branch for now)

File: src/filter/macros.rs (hunk 15 above)
Comment (HadrienG2) [reply src/filter/macros.rs:51]: Exactly.

Comment (VanillaBrooks) [reply src/filter/macros.rs:51]: I was originally using `get_unchecked` but instead used `.get().expect()` to see where my indexing was incorrect in the for loop. Since `get_unchecked` has less overhead ill be sure to switch it back.

File: src/error.rs (hunk 1 above)
Comment (VanillaBrooks) [reply src/error.rs:62]: It does seem like one of the more questionable language design choices, but seeing as the following code compiles:

```
// main.rs
use error::{Error, MatrixError};

fn coerce_to_Error() -> Result<(), Error>{
    // ? operator calls Error::from(_) on error variant
    // which the macro implemented
    let ok_value = return_MatrixError()?;
    return Ok(ok_value)
}

fn return_MatrixError() -> Result<() , MatrixError> {
    return Err(MatrixError::NonInvertible)
}
```

It should not be a problem

File: src/filter/filter_gain.rs (hunk 5 above)
Comment (VanillaBrooks) [reply src/filter/filter_gain.rs]: I dont see anything mentioning `nalgebra`'s `MatrixMN` being threadsafe in the [docs](https://docs.rs/nalgebra/0.18.0/nalgebra/base/type.MatrixMN.html), but seeing as this compiles I think we are in good territory:

```
use config::*;
fn test_threadsafe<T: Send + Sync>(input: &T) {}

fn construct_identity() {
    lazy_static!{
        static ref identity: Mat4 = Mat4::identity();
    }
    test_threadsafe(&identity);
}
```

As for compile time optimization, I had unexpected runtime errors with `nagebra` at some points, so I was hesitant to trust that this would be compiled away. This comment is more of a reminder to make sure that `identity` becomes static when compiled.

File: src/filter/utils.rs (hunk 11 above)
Comment (VanillaBrooks) [reply src/filter/utils.rs:19]: This code was only used to quickly construct some test values for the matricies and vectors so I could pass them into `filter::linear::run` and check for runtime errors. Its not meant for any sort of practical use. Will comment it to reflect this.

File: src/filter/filter_gain.rs (hunk 7 above)
Comment (VanillaBrooks) [reply src/filter/filter_gain.rs:57]: I just tried to keep the naming somewhat consistent between `prediction.rs` `filter_gain.rs` `filter_mean.rs` and `smoothing.rs` and parallel to the paper's description. 
I thought it might make sense to see not name it `filtered_residual_matrix` since it is already in `filter_gain.rs`. How would you name it?

File: src/error.rs (hunk 1 above)
Comment (paulgessinger) [reply src/error.rs:62]: Seems fair. I was just curious.

File: src/filter/filter_gain.rs (hunk 5 above)
Comment (paulgessinger) [reply src/filter/filter_gain.rs]: Ok, sure. `nalgebra`'s matrices are not thread-safe in all likelihood. I was just wondering if a lazily initialized static would be thread safe in general, not specifically the matrix identity here.

File: src/filter/prediction.rs (hunk 13 above)
Comment (VanillaBrooks) [reply src/filter/prediction.rs]: Not sure what you mean. There is a `residual_mat` function in `filter_gain.rs` `smoothing.rs` and `prediction.rs`. Each one calculates the residual with respect to the module they are in. The `residual_mat` function in `filter_gain.rs` and `prediction.rs` are not the same code. 

This was an attempt to keep things consistent. I thought calling `filter_gain::residual_mat(_, _, _)` and `prediction::residual_mat(_, _, _)` would be self-explanatory as to what they achieve (as opposed to `filter_gain::filtered_residual_mat(_, _, _)` and `prediction::pred_residual_mat(_, _, _)` ) What do you think ?

File: src/filter/filter_gain.rs (hunk 5 above)
Comment (HadrienG2) [reply src/filter/filter_gain.rs]: Yup, lazy_static works more or less like static function var initialization in C++ and Java : initialize on first access, block other threads during the initialization process, and then every subsequent access has to do an atomic read and branch to check for initialization. Unlike in those languages, however, there is a requirement of thread-safe accesses (e.g. only reads) after initialization.

It was initially meant as an alternative to introducing "life before main" in Rust. But these days, many of its uses are being replaced by `const fn` compile-time constructors, which are basically Rust's take on constexpr methods.

Comment (paulgessinger) [reply src/filter/filter_gain.rs]: 👍

**Comment by paulgessinger** [comment]
Should we try and merge this soon?

File: src/geometry/rectangle.rs (hunk 24 above)
Comment (VanillaBrooks) [reply src/geometry/rectangle.rs]: Added an epsilon of 0.0005

File: src/geometry/rectangle.rs (hunk 22 above)
Comment (VanillaBrooks) [reply src/geometry/rectangle.rs]: changed:
`contains_from_local()` => `inside()`
`contains_from_global()` => `inside_global()`

File: src/geometry/rectangle.rs (hunk 20 above)
Comment (VanillaBrooks) [reply src/geometry/rectangle.rs]: updated comment

File: src/filter/filter_gain.rs (hunk 4 above)
Comment (VanillaBrooks) [reply src/filter/filter_gain.rs]: updated all the single letter variable names to better descriptions (and more consistent) in `filter_gain.rs`, `filter_means.rs` `prediction.rs` and `smoothing.rs`

File: src/filter/filter_gain.rs (hunk 3 above)
Comment (VanillaBrooks) [reply src/filter/filter_gain.rs]: its not. changed

File: src/filter/macros.rs (hunk 16 above)
Comment (VanillaBrooks) [reply src/filter/macros.rs:71]: I _think_ I mean iterate pairwise. The idea is that once we move to the next sensor the `$current` variable would become the `$previous`, the `$next` became `$current`, and so on. 
This macro was used heavily when originally working with iterators since there was a lot of repetitive code in the smoothing phase. I left it in since i expect to be working with iterators in the future. 
The replacement to this (which gives more control) is `get_unchecked!`, which _should_ have the same performance as the older iterator code i was working with.

File: src/filter/prediction.rs (hunk 14 above)
Comment (VanillaBrooks) [reply src/filter/prediction.rs:35]: A little confused on this. Are you saying that `residual_vec()` wont work in non-linear cases?

File: src/filter/prediction.rs (hunk 12 above)
Comment (VanillaBrooks) [reply src/filter/prediction.rs]: right. I also added a `linear_state_vector()` to predict the current state vector based on the current phi / theta projected to the next sensor.

File: src/geometry/rectangle.rs (hunk 21 above)
Comment (VanillaBrooks) [reply src/geometry/rectangle.rs]: changed

File: src/filter/linear.rs (hunk 18 above)
Comment (VanillaBrooks) [reply src/filter/linear.rs]: I hard coded that matrix into `linear::run`. makes sense that they are all the same matrix. thanks

File: src/filter/linear.rs (hunk 17 above)
Comment (VanillaBrooks) [reply src/filter/linear.rs:20]: added

File: src/filter/filter_means.rs (hunk 9 above)
Comment (VanillaBrooks) [reply src/filter/filter_means.rs]: changed for consistency. thanks

File: src/filter/filter_gain.rs (hunk 2 above)
Comment (VanillaBrooks) [reply src/filter/filter_gain.rs]: changed the variable name to `pred_state_vec`. added more descriptive variable names to all the KF equations.

File: src/filter/filter_gain.rs (hunk 6 above)
Comment (VanillaBrooks) [reply src/filter/filter_gain.rs]: fixed.

File: src/filter/filter_gain.rs (hunk 8 above)
Comment (VanillaBrooks) [reply src/filter/filter_gain.rs]: `residual_vec` does `sensor_mapping_mat * kalman_gain_mat` whereas `state_vector` does `sensor_mapping_mat * pred_state_vec`. Also, `residual_vec` has an extra identity matrix in the calculation. 

Can you elaborate on what you mean?

**Comment by VanillaBrooks** [comment]
Should be good to merge. All tests pass. A few lingering conversations to follow up with but nothing critical I believe.

File: src/filter/filter_gain.rs (hunk 7 above)
Comment (paulgessinger) [reply src/filter/filter_gain.rs:57]: I see what you mean. I don't have an immediate suggestion on what to call it instead, so let's keep it for now.

File: src/filter/filter_gain.rs (hunk 8 above)
Comment (paulgessinger) [reply src/filter/filter_gain.rs]: You're right, it's different.

File: src/filter/prediction.rs (hunk 13 above)
Comment (paulgessinger) [reply src/filter/prediction.rs]: Ok, I see what you're going for here.

File: src/filter/prediction.rs (hunk 14 above)
Comment (paulgessinger) [reply src/filter/prediction.rs:35]: No, it should. I was talking about the duplication, but it is in fact different.

File: src/filter/macros.rs (hunk 15 above)
Comment (paulgessinger) [reply src/filter/macros.rs:51]: Ok
